### DISCUSSION TITLE
writing generation metadata into files

### DISCUSSION THREAD

**Discussion by DasWookie** [hlky/stable-diffusion-webui#680] [Feature suggestions]
It would be terrific in the file generation inputs were written into the files metadata as properties instead of (or in addition to) a separate YAML file. That way, for img2img or imagelab the inputs could be loaded as default inputs, it would be easy to examine an image to determine what inputs were used to create it, as well as making it easier to share the values used to generate the file when the image is posted in case you lost track of or can't remember what was used as the generation inputs.

You can store metadata in Pillow using the PngImagePlugin.PngInfo library. See https://pillow.readthedocs.io/en/stable/PIL.html#PIL.PngImagePlugin.PngInfo for usage, but here is an example implementation that shows how to add properties and values to an image:
from PIL import Image
from PIL.PngImagePlugin import PngInfo
targetImage = Image.open("Sample.png")
metadata = PngInfo()
metadata.add_text("Prompt", "A corgi wearing a tophat as an oil painting")
metadata.add_text("Height", str(512))
metadata.add_text("Width", str(768))
metadata.add_text("Scale", str(7.5))
metadata.add_text("Seed", str(8675309))
metadata.add_text("Steps", str(45))
metadata.add_text("Diffuser", "k_euler")
metadata.add_text("Prompt Weights", str(1))
metadata.add_text("GPFGAN", "True")
metadata.add_text("ERSGAN", "True")
targetImage.save("NewSample.png", pnginfo=metadata)
targetImage = Image.open("NewSample.png")
print(targetImage.text)
targetImage.close()

Example run showing the info which was saved into a file:
(lsd) [cyber@Highside scripts]$ python attributes.py
{'Prompt': 'A corgi wearing a tophat as an oil painting', 'Height': '512', 'Width': '768', 'Scale': '7.5', 'Seed': '8675309', 'Steps': '45', 'Diffuser': 'k_euler', 'Prompt Weights': '1', 'GPFGAN': 'True', 'ERSGAN': 'True'}

There's an exif module as well, which would allow you to see the information easily in windows explorer

**Comment by EliEron** [comment]
This is actually already supported. It was implemented in PR #78. But you need to add the argument "--save-metadata" to the additional_arguments option in the relauncher.py script.

It doesn't implement all of the metadata you mention, but it implements: height, width, cfg_scale, prompt, seed, steps, GFPGAN, normalize_prompt_weights and denoising_strength (Only for img2img).

**Comment by DasWookie** [comment]
Doesn't appear to be merged into master yet. I reverified I'm on the main branch, and I don't see the changes. Looking at the commit history, it looks like it got reverted out so... looks like it isn't implemented anymore. {shrug}

**Reply by EliEron** [reply]
No, it's definitively still merged. I just generated this image on the latest master:
![00009-32_k_euler_a_2809143907_0 00](https://user-images.githubusercontent.com/22575945/188523721-72dc0cb0-7e21-4cef-82b2-9e7b11695bf4.png)

And if you download it and look at its metadata you'll see all of the info is there.

I don't see anything in the commit log that mentions reverting it. Can you elaborate on where you saw the revert?

**Comment by DasWookie** [comment]
Ah, right you are. I was looking at the wrong place and can confirm the change is there for png files. Doesn't work for JPG, which sucks as that's what I use to save on disk space, but it's a start.
